block/submit.go

```diff
@@ -2,200 +2,210 @@ package block
 
 import (
 	"context"
+	"errors"
 	"fmt"
+	"sync/atomic"
 	"time"
 
-	"github.com/dymensionxyz/gerr-cosmos/gerrc"
-
 	"github.com/dymensionxyz/dymint/da"
-	"github.com/dymensionxyz/dymint/node/events"
+	"github.com/dymensionxyz/dymint/store"
 	"github.com/dymensionxyz/dymint/types"
-	uevent "github.com/dymensionxyz/dymint/utils/event"
+	uchannel "github.com/dymensionxyz/dymint/utils/channel"
+	"github.com/dymensionxyz/gerr-cosmos/gerrc"
+	"golang.org/x/sync/errgroup"
 )
 
 // SubmitLoop is the main loop for submitting blocks to the DA and SL layers.
 // It submits a batch when either
 // 1) It accumulates enough block data, so it's necessary to submit a batch to avoid exceeding the max size
 // 2) Enough time passed since the last submitted batch, so it's necessary to submit a batch to avoid exceeding the max time
-func (m *Manager) SubmitLoop(ctx context.Context) (err error) {
-	maxTime := time.NewTicker(m.Conf.BatchSubmitMaxTime)
-	defer maxTime.Stop()
-
-	// get produced size from the block production loop and signal to submit the batch when batch size reached
-	maxSizeC := make(chan struct{}, m.Conf.MaxSupportedBatchSkew)
-	go m.AccumulatedDataLoop(ctx, maxSizeC)
-
-	// defer func to clear the channels to release blocked goroutines on shutdown
-	defer func() {
-		m.logger.Info("Stopped submit loop.")
+// It will back pressure (pause) block production if it falls too far behind.
+func (m *Manager) SubmitLoop(ctx context.Context,
+	bytesProduced chan int,
+) (err error) {
+	return SubmitLoopInner(ctx,
+		bytesProduced,
+		m.Conf.MaxBatchSkew,
+		m.Conf.BatchSubmitMaxTime,
+		m.Conf.BatchMaxSizeBytes,
+		m.CreateAndSubmitBatchGetSizeEstimate,
+	)
+}
 
+// SubmitLoopInner is a unit testable impl of SubmitLoop
+func SubmitLoopInner(ctx context.Context,
+	bytesProduced chan int, // a channel of block and commit bytes produced
+	maxBatchSkew uint64, // max number of batches that submitter is allowed to have pending
+	maxBatchTime time.Duration, // max time to allow between batches
+	maxBatchBytes uint64, // max size of serialised batch in bytes
+	createAndSubmitBatchGetSizeEstimate func(maxSize uint64) (uint64, error),
+) error {
+	eg, ctx := errgroup.WithContext(ctx)
+
+	pendingBytes := atomic.Uint64{}
+	counter := uchannel.NewNudger()   // used to avoid busy waiting (using cpu) on counter thread
+	submitter := uchannel.NewNudger() // used to avoid busy waiting (using cpu) on submitter thread
+
+	eg.Go(func() error {
```

Comment:
>encapsulte in seperate funcs (counter, submitter)

IMO in this case it would not be more readable, as it is written it's quite easy to see all the logic one screen page, and there is no advantage in testing from splitting it out. Also, they both use the three variables pendingBytes, trigger, submitter, so they'd need to be passed as args

> IMO SubmitLoopInner can be merged back with SubmitLoop

In this case I disagree.  SubmitLoopInner  is very testable
It's a pretty common pattern to have an inner impl with testable DI and a wrapper (SubmitLoop) to use defaults